apps/governance/src/components/web3-connector/web3-connector.tsx

```diff
@@ -52,23 +49,61 @@ export function Web3Connector({
 interface Web3ContentProps {
   children: ReactElement;
   appChainId: number;
-  setDialogOpen: (isOpen: boolean) => void;
 }
 
 export const Web3Content = ({ children, appChainId }: Web3ContentProps) => {
+  const { appState, appDispatch } = useAppState();
   const { connector, chainId } = useWeb3React();
+  const [previousChainId, setPreviousChainId] = useState(chainId);
   const error = useWeb3ConnectStore((store) => store.error);
   const disconnect = useWeb3Disconnect(connector);
 
+  const showDisconnectNotice = useCallback(
+    (isVisible: boolean) =>
+      appDispatch({
+        type: AppStateActionType.SET_DISCONNECT_NOTICE,
+        isVisible,
+      }),
+    [appDispatch]
+  );
+
   useEffect(() => {
     if (connector?.connectEagerly) {
       connector.connectEagerly();
     }
-    // wallet connect doesnt handle connectEagerly being called when connector is also in the
+    // wallet connect doesn't handle connectEagerly being called when connector is also in the
     // deps array.
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, []);
 
+  useEffect(() => {
+    if (chainId !== undefined) {
+      // We use this to detect when the user switches networks.
+      setPreviousChainId(chainId);
+
+      if (chainId !== appChainId) {
+        disconnect();
+
+        // If the user was previously connected, show the disconnect explanation notice.
+        if (previousChainId !== undefined && !appState.disconnectNotice) {
```

Comment:
question (blocking): `previousChainId !== undefined` is always true because of if on line 80.

Reply:
Talked through on slack

Reply:
Discussed in DMs, but essentially:

- It's possible for previousChainId to be undefined whilst the current chain id is not undefined
- If I remove that 'previousChainId !== undefined' check, and if a user still has a metamask connection and they're on the wrong network, we show the user that they've been disconnected all the time. This is true, but it's needlessly noisy. I'm aiming to only show them the disconnected notice after they deliberately change network. Whether this is the right thing to do is up for debate